presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveCrossJoinWithConstantInput.java

```diff
@@ -0,0 +1,145 @@
+            joinInput = context.getLookup().resolve(node.getRight());
+        }
+        else {
+            return Result.empty();
+        }
+        Optional<Map<VariableReferenceExpression, RowExpression>> mapping = getConstantAssignments(singleValueInput, context);
+        if (!mapping.isPresent()) {
+            return Result.empty();
+        }
+        PlanNode resultNode = addProjections(joinInput, context.getIdAllocator(), mapping.get());
+        if (node.getFilter().isPresent()) {
+            resultNode = new FilterNode(node.getSourceLocation(), context.getIdAllocator().getNextId(), resultNode, node.getFilter().get());
+        }
+        return Result.ofPlanNode(resultNode);
+    }
+
+    private boolean isOutputSingleConstantRow(PlanNode planNode, Context context)
+    {
+        while (planNode instanceof ProjectNode) {
+            planNode = context.getLookup().resolve(((ProjectNode) planNode).getSource());
+        }
+        if (planNode instanceof ValuesNode) {
+            return ((ValuesNode) planNode).getRows().size() == 1;
+        }
+        return false;
+    }
+
+    private Optional<Map<VariableReferenceExpression, RowExpression>> getConstantAssignments(PlanNode planNode, Context context)
+    {
+        List<VariableReferenceExpression> outputVariables = planNode.getOutputVariables();
+        Map<VariableReferenceExpression, RowExpression> mapping = outputVariables.stream().collect(toImmutableMap(Function.identity(), Function.identity()));
+        while (planNode instanceof ProjectNode) {
+            Map<VariableReferenceExpression, RowExpression> assignments = ((ProjectNode) planNode).getAssignments().getMap();
+            mapping = mapping.entrySet().stream().collect(toImmutableMap(
+                    Map.Entry::getKey, entry -> entry.getValue() instanceof VariableReferenceExpression ? assignments.get(entry.getValue()) : entry.getValue()));
+            planNode = context.getLookup().resolve(((ProjectNode) planNode).getSource());
+        }
+        checkState(planNode instanceof ValuesNode);
+        ValuesNode valuesNode = (ValuesNode) planNode;
+        if (!valuesNode.getOutputVariables().isEmpty()) {
+            Map<VariableReferenceExpression, RowExpression> assignments = IntStream.range(0, valuesNode.getOutputVariables().size()).boxed()
+                    .collect(toImmutableMap(idx -> valuesNode.getOutputVariables().get(idx), idx -> valuesNode.getRows().get(0).get(idx)));
+            mapping = mapping.entrySet().stream().collect(toImmutableMap(
+                    Map.Entry::getKey, entry -> entry.getValue() instanceof VariableReferenceExpression ? assignments.get(entry.getValue()) : entry.getValue()));
+        }
+        boolean allDeterministic = mapping.entrySet().stream().allMatch(entry -> rowExpressionDeterminismEvaluator.isDeterministic(entry.getValue()));
+        if (allDeterministic) {
+            return Optional.of(mapping);
+        }
+        return Optional.empty();
+    }
```

Comment:
some nits :
1. some opportunities to avoid code duplication

these lines are repeated twice and can be refactored to a method if it makes sense.
`entry -> entry.getValue() instanceof VariableReferenceExpression ? assignments.get(entry.getValue()) : entry.getValue()`

`mapping.entrySet().stream().collect(toImmutableMap(
                    Map.Entry::getKey, entry -> entry.getValue() instanceof VariableReferenceExpression ? assignments.get(entry.getValue()) : entry.getValue()));`
                    
 2. Add a line break before `checkState` for readability.
 
 3. 
 since we are operating on values we don't have to reach out to entrySet and then get the value. Instead directly call values() .
 
 `boolean allDeterministic = mapping.entrySet().stream().allMatch(entry -> rowExpressionDeterminismEvaluator.isDeterministic(entry.getValue()));`
->
 `boolean allDeterministic = mapping.values().stream().allMatch(rowExpressionDeterminismEvaluator::isDeterministic);`